pkg/cluster/client_builder.go

```diff
@@ -39,16 +42,34 @@ func NewClientBuilder() ClientBuilder {
 }
 
 type newClientBuilder struct {
-	uncached []client.Object
+	uncached      []client.Object
+	newClientFunc NewClientFunc
 }
 
 func (n *newClientBuilder) WithUncached(objs ...client.Object) ClientBuilder {
 	n.uncached = append(n.uncached, objs...)
 	return n
 }
 
+func (n *newClientBuilder) WithNewClientFunc(f NewClientFunc) ClientBuilder {
+	n.newClientFunc = f
+	return n
+}
+
 func (n *newClientBuilder) Build(cache cache.Cache, config *rest.Config, options client.Options) (client.Client, error) {
-	// Create the Client for Write operations.
+	if n.newClientFunc == nil {
+		n.newClientFunc = DefaultNewClientFunc
+	}
+
+	return n.newClientFunc(cache, config, options, n.uncached...)
+}
+
+// NewClientFunc is the signature of the func that is called to construct a new client.
+type NewClientFunc func(cache cache.Cache, config *rest.Config, options client.Options, uncachedObjects ...client.Object) (client.Client, error)
```

Comment:
@vincepri on second thought at this point where we have to export NewClienFunc again (with a slightly different signature than what it had before) it might be easier to simply remove the ClientBuilder and go back to using a NewClientFunc directly. WDYT?

Reply:
Could we instead create a builder from a function?

Reply:
> Could we instead create a builder from a function?

Can you elaborate, I don't think I understand what you mean by that?